### PR TITLE
filter: check if connection is open before and after filter read/write

### PR DESCRIPTION
**Pull request by crazyxy**
Signed-off-by: Yan Xue <yxyan@google.com>

Description: Check if connection is open before and after filter read/write. In `onData` or `onWrite`, filter could close the underlying connection.
Risk Level: Low
Testing: unit test
Docs Changes: N/A
Release Notes: N/A
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by crazyxy** [comment]
> Thanks for working on this. While I think this change is generally correct, historically we have assumed that filters will stop iteration if they close the connection/stream. WDYT about just putting in asserts to this effect? it would be a much less scary change.
> 
> /wait-any

I think the assumption is totally correct but sometimes users might not follow it or they are not intend to break the assumption. I found this bug from a filter extended by istio/mixer. 

I would prefer providing some tolerance when users don't follow the convention rather than terminating the process.

File: source/common/network/filter_manager_impl.cc [hunk 1]
```diff
@@ -81,10 +82,13 @@ FilterStatus FilterManagerImpl::onWrite(ActiveWriteFilter* filter,
   }
 
   for (; entry != downstream_filters_.end(); entry++) {
-    StreamBuffer write_buffer = buffer_source.getWriteBuffer();
-    FilterStatus status = (*entry)->filter_->onWrite(write_buffer.buffer, write_buffer.end_stream);
-    if (status == FilterStatus::StopIteration) {
-      return status;
+    if (connection_.state() == Connection::State::Open) {
```
Comment (jplevyak) [inline source/common/network/filter_manager_impl.cc]: You can pull this outside the loop.

**Comment by mattklein123** [comment]
> I think either of them is ok. I would prefer providing some tolerance when users don't follow the convention rather than terminating the process. WDYT?

Fair enough. Sounds good.

File: source/common/network/filter_manager_impl.cc [hunk 2]
```diff
@@ -46,18 +46,19 @@ void FilterManagerImpl::onContinueReading(ActiveReadFilter* filter,
   }
 
   for (; entry != upstream_filters_.end(); entry++) {
-    if (!(*entry)->initialized_) {
+    if (!(*entry)->initialized_ && connection_.state() == Connection::State::Open) {
```
Comment (mattklein123) [inline source/common/network/filter_manager_impl.cc]: Do you need this check with the return below? It seems like this is a straight up bug that would be better served via an ASSERT?

File: source/common/network/filter_manager_impl.cc [hunk 3]
```diff
@@ -46,18 +46,19 @@ void FilterManagerImpl::onContinueReading(ActiveReadFilter* filter,
...
-      if (status == FilterStatus::StopIteration) {
+      if (status == FilterStatus::StopIteration || connection_.state() != Connection::State::Open) {
         return;
       }
     }
 
     StreamBuffer read_buffer = buffer_source.getReadBuffer();
-    if (read_buffer.buffer.length() > 0 || read_buffer.end_stream) {
+    if ((read_buffer.buffer.length() > 0 || read_buffer.end_stream) &&
+        connection_.state() == Connection::State::Open) {
```
Comment (mattklein123) [inline source/common/network/filter_manager_impl.cc]: Same here, shouldn't we only be checking after a filter runs if that is what we want to do?

File: source/common/network/filter_manager_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/network/filter_manager_impl.cc]: Again, prefer a check only after a filter runs.

File: source/common/network/filter_manager_impl.cc (hunk 2 above)
Comment (crazyxy) [reply source/common/network/filter_manager_impl.cc]: I think we might need this check here. For example, there could be some filter which returns `FilterStatus::StopIteration` immediately, closes the connection and uses callback to call this method to continue iteration.

Comment (mattklein123) [reply source/common/network/filter_manager_impl.cc]: Then can't you have a single check at the top of the function? Please add a test for this case specifically if the test doesn't cover it.

Comment (crazyxy) [reply source/common/network/filter_manager_impl.cc]: I will move the check to the top of the function.

File: source/common/network/filter_manager_impl.cc [hunk 4]
```diff
@@ -73,6 +79,12 @@ FilterStatus FilterManagerImpl::onWrite() { return onWrite(nullptr, connection_)
 
 FilterStatus FilterManagerImpl::onWrite(ActiveWriteFilter* filter,
                                         WriteBufferSource& buffer_source) {
+  // Filter could return status == FilterStatus::StopIteration immediately, close the connection and
+  // use callback to call this function.
+  if (connection_.state() != Connection::State::Open) {
```
Comment (mattklein123) [inline source/common/network/filter_manager_impl.cc]: Don't you still need the check below after calling the filter? If we are going to do these we should be complete for all cases?

Comment (crazyxy) [reply source/common/network/filter_manager_impl.cc]: Good catch. We need to check after calling the filter.

**Comment by crazyxy** [comment]
> Thanks, do you have test coverage for all of these new cases?
> 
> /wait

I have 2 test which should cover these cases.

**Comment by mattklein123** [comment]
If you were able to change the if statement after the filter call without changing your tests, I don't think your tests are complete. Please make them more robust to cover all cases.

/wait

**Comment by crazyxy** [comment]
> If you were able to change the if statement after the filter call without changing your tests, I don't think your tests are complete. Please make them more robust to cover all cases.
> 
> /wait

I didn't add test for `onWrite`. Now the test case should cover both read and write. @mattklein123 could you please take a look?

**Pull request merged**